rct229/rulesets/ashrae9012019/ruleset_functions/get_BPF_building_area_types_and_zones.py

```diff
@@ -0,0 +1,124 @@
+from typing import TypedDict
+
+from pint import Quantity
+from rct229.rulesets.ashrae9012019.data_fns.table_lighting_space_type_BPF_area_type_map import (
+    lighting_space_type_to_BPF_area_type,
+)
+from rct229.rulesets.ashrae9012019.ruleset_functions.get_zone_BPF_BAT import (
+    get_zone_BPF_BAT,
+)
+from rct229.utils.jsonpath_utils import find_all
+from rct229.utils.pint_utils import ZERO
+
+
+class ClassificationType:
+    BUILDING_SEGMENT_LIGHTING: str = "BUILDING_SEGMENT_LIGHTING"
+    SPACE_LIGHTING: str = "SPACE_LIGHTING"
+
+
+class BuildingAreaTypeWithTotalAreaAndZone(TypedDict):
+    zone_id: list[str]
+    area: Quantity
+    classification_source: str
+
+
+def get_BPF_building_area_types_and_zones(
+    rmd: dict,
+) -> dict[str, BuildingAreaTypeWithTotalAreaAndZone]:
+    """
+    Get a dictionary of the zone_ids associated with each BPF building area type associated with U_RMD, P_RMD, or B_RMD. Also returns the total floor area of each building area type and the source of the information (BUILDING_AREA_LIGHTING or SPACE_LIGHTING)
+
+    Parameters
+    ----------
+    rmd: str, The ruleset model descriptions
+
+    Returns
+    -------
+    building_area_types_with_total_area_and_zones_dict: A dict that saves all the BPF building area types and includes a list of all the zone ids associated with area type as well as
+                                                       the total area of each building area type: {MULTIFAMILY: {"zone_id": ["zone_2", "zone_3", "zone_4"], "area": 34567, "classification_source": "BUILDING_SEGMENT_LIGHTING"}, HEALTHCARE_HOSPITAL: {"zone_id": ["zone_1","zone_5","zone_6"], "area": 20381, "classification_source": "SPACE_LIGHTING"}
+    """
+
+    def _define_bldg_segment_BPF_BAT(
+        building_area_types_with_total_area_and_zones_dict: dict,
+        classification_source: str,
+    ) -> dict:
+
+        building_area_types_with_total_area_and_zones_dict[building_segment_BPF_BAT] = {
+            "zone_id": [],
+            "area": ZERO.AREA,
+            "classification_source": classification_source,
+        }
+
+        return building_area_types_with_total_area_and_zones_dict
```

Comment:
This function is not a best practice in functioning programming concept.
It passes in a variable, modifies the variable, and returns the same variable.
It is not ideal from the memory pointer view and could cause memory leaks. But in Python, the memory is protected so it is OK.

If you want something like this, consider a function just simply initialize a dictionary such as:
```python
def _initialize_bldg_segment_bpf_bat() -> dict:
    return {
        "zone_id": [],
        "area": ZERO.AREA
        "classification_source: ClassificationType. BUILDING_SEGMENT_LIGHTING
    }
```
Then in the body, you can modify the parameters accordingly.

Reply:
Good point and thank you for letting me know about this! 

I removed the ```_initialize_bldg_segment_bpf_bat``` function and defined the ```building_area_types_with_total_area_and_zones_dict``` below. The main reason why I created the function was I wanted to get rid of redundancy.